### PR TITLE
Export all local exported variables in a new scope

### LINKED ISSUE #6153: local-exported variables not correctly exported when another block is in-between
If I override `PATH` as a local variable, `type` respects this if the local scope is the current scope, but not if it's an enclosing scope.

For the following code, assume `/opt/brew/bin/brew` exists on disk but isn't in my normal `PATH`:

```
> type brew
type: Could not find 'brew'
> set -lxp PATH /opt/brew/bin
> type brew
brew is /opt/brew/bin/brew
> begin; type brew; end
type: Could not find 'brew'
```

I first ran into this when `eval type brew` didn't work but `type brew` did, which is explained by this bug as `eval` establishes a nested scope.

fish, version 3.0.2

### PR DESCRIPTION
**Pull request by krobelus**
Fixes #6153

----

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: src/env.cpp [hunk 1]
```diff
@@ -936,13 +936,16 @@ void env_stack_impl_t::push_nonshadowing() {
...
 void env_stack_impl_t::push_shadowing() {
     // Propagate local exported variables.
-    // TODO: this should take all local exported variables, not just those in the top scope.
     auto node = std::make_shared<env_node_t>(true, nullptr);
-    for (const auto &var : locals_->env) {
-        if (var.second.exports()) {
-            node->env.insert(var);
-            node->changed_exported();
+    auto cursor = locals_;
+    while (cursor) {
```
Comment (faho) [inline src/env.cpp]: Why is this not a `for` loop as well? Would save on the manual `cursor = cursor->next`.

Comment (krobelus) [reply src/env.cpp]: oh thanks, changed

Comment (faho) [reply src/env.cpp]: Did you forget to push? I'm still seeing the while-loop?

Comment (krobelus) [reply src/env.cpp]: yeah my bad, I pushed it now

**Comment by ridiculousfish** [comment]
Merged as 9b86d5dd16e99dd5d785ee6263c42af4c6932726, thank you!

**Pull request closed without merging**